# bioinfo_warehouse


- [bioinfo_warehouse](#bioinfo_warehouse)
    - [introduction](#introduction)
        - [pair_align_dna](#pair_align_dna)
        - [mystring](#mystring)

## introduction
The bioinfo_warehouse stores codes about bioinformational algorithms. Now it includes sequence alignment. Soon i will update the code and implement my algorithms. This markdown will be updated and introduce the details of algorithms.
As we know, the decreasing cost of DNA sequence causes the development of bioinformation. We are able to get the information of DNA sequences in specific organisms because the technology we use is cheap. Of course, with more and more data that we sequence, we need to store, integrate inforamtion and we also need to search and analyze data in a short time. Here is why bioinformation comes. 
One thing we need to pay attention to is that the main research target in this field is **sequences**, the sequences of DNA, RNA or numerous proteins, so most of algorithms in this field are about how to deal with them. This is a restriction for bioinfo scientists
. 
# sequence  

## pair_align_dna
This algorithm is used for searching **similiar sequences**, which can help us to design the primer in PCR or find the similiar genes between organisms. Of course, it can be used to find motifs in DNA sequence, but I will introduce other several ways to find for they can find motifs without knowing the DNA sequence.

# similiar sequences


## mystring
I don't want to use string class in C++, because I want to insert some functions. 
